src/devices/bus/spectrum/plusd.cpp

```diff
@@ -0,0 +1,374 @@
+	
+	ROM_REGION( 0x200, "plds", 0 )
+	ROM_LOAD( "alice_pal20l8.ic4", 0x000, 0x144, CRC(60135856) SHA1(41273f13a3680b29ba84ae1e85829482c783c55e) )
+ROM_END
+
+//-------------------------------------------------
+//  device_add_mconfig - add device configuration
+//-------------------------------------------------
+
+void spectrum_plusd_device::device_add_mconfig(machine_config &config)
+{
+	WD1772(config, m_fdc, 16_MHz_XTAL / 2);
+	FLOPPY_CONNECTOR(config, "fdc:0", plusd_floppies, "35dd", spectrum_plusd_device::floppy_formats).enable_sound(true);
+	FLOPPY_CONNECTOR(config, "fdc:1", plusd_floppies, "35dd", spectrum_plusd_device::floppy_formats).enable_sound(true);
+	
+	/* printer port */
+	CENTRONICS(config, m_centronics, centronics_devices, "printer");
+	m_centronics->busy_handler().set(FUNC(spectrum_plusd_device::busy_w));
+	
+	/* software list */
+	SOFTWARE_LIST(config, "flop_list").set_original("spectrum_mgt_flop");
+}
+
+
+const tiny_rom_entry *spectrum_plusd_device::device_rom_region() const
+{
+	return ROM_NAME(plusd);
+}
+
+
+//**************************************************************************
+//  LIVE DEVICE
+//**************************************************************************
+
+//-------------------------------------------------
+//  spectrum_plusd_device - constructor
+//-------------------------------------------------
+
+spectrum_plusd_device::spectrum_plusd_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, SPECTRUM_PLUSD, tag, owner, clock)
+	, device_spectrum_expansion_interface(mconfig, *this)
+	, m_rom(*this, "rom")
+	, m_fdc(*this, "fdc")
+	, m_floppy(*this, "fdc:%u", 0)
+	, m_centronics(*this, "centronics")
+{
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void spectrum_plusd_device::device_start()
+{
+	memset(m_ram, 0, sizeof(m_ram));
+
+	save_item(NAME(m_romcs));
+	save_item(NAME(m_ram));
+	save_item(NAME(m_centronics_busy));
+}
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+
+void spectrum_plusd_device::device_reset()
+{
+	m_romcs = 0;
+	m_centronics_busy = false;
+}
+
+
+//**************************************************************************
+//  IMPLEMENTATION
+//**************************************************************************
+
+READ_LINE_MEMBER(spectrum_plusd_device::romcs)
+{
+	return m_romcs;
+}
+
+void spectrum_plusd_device::pre_opcode_fetch(offs_t offset)
+{
+	if (!machine().side_effects_disabled())
+	{
+		switch (offset)
+		{
+		case 0x0000:
+		case 0x0008:
+		case 0x003a:
+		case 0x0066:
+			m_romcs = 1;
+			break;
+		}
+	}
+}
+
+uint8_t spectrum_plusd_device::iorq_r(offs_t offset)
+{
+	uint8_t data = 0xff;
+
+	if (!machine().side_effects_disabled())
```

Comment:
The only thing here with direct side effects is setting `m_romcs`.  If reading the FDC has side effects it should be guarding itself.  Wrapping the entire body like this makes debugging more difficult (e.g. you can't see Centronics busy status).  You should only guard the thing with direct side effects.

Reply:
I must admit I only included this as I saw it used in the other slot devices and assumed it must be required for some reason. I'm not sure exactly what "side effects" are or why I should need to turn the off. Could you explain or point me to a source?

Reply:
Side effects are reads from a source other than the CPU, notably the memory display in the debugger.

Reply:
I see, debugger guard, thanks, I will make some changes.